### PR TITLE
asset-swapper: Fix optimization of buy paths

### PR DESCRIPTION
**Pull request by dorothy-zbornak**
## Description

The adjusted rate for incomplete buy paths were being calculated wrongly, which caused us to perform unnecessary and detrimental splits on certain paths (e.g., DAI->BAL).

Also some other tweaks and significant speed ups to the path optimizer.

#### Buys and sells over ETH, DAI, USDC, USDT, WBTC, BAL against prod

realized-eth-dai-usdc-usdt-wbtc-bal

#### Latency over `development`
⚠ *colors have flipped* ⚠
latency

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts [hunk 1]
```diff
@@ -122,16 +146,7 @@ function clipFillAdjustedOutput(fill: Fill, remainingInput: BigNumber): BigNumbe
...
-
-function getRate(side: MarketOperation, input: BigNumber, output: BigNumber): BigNumber {
-    if (input.eq(0) || output.eq(0)) {
-        return ZERO_AMOUNT;
-    }
-    if (side === MarketOperation.Sell) {
-        return output.div(input);
-    }
-    return input.div(output);
+    return remainingInput.times(fill.output.div(fill.input)).plus(penalty);
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts:151]: Bug fix. `fill.rate != fill.output / fill.input`

File: packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts [hunk 2]
```diff
@@ -55,64 +66,77 @@ function mixPaths(
...
+        return a.index - b.index;
+    });
+    _walk([], ZERO_AMOUNT, ZERO_AMOUNT, 0, sortedFills);
+    if (!isValidPath(bestPath)) {
+        throw new Error('nooope');
+    }
     return bestPath;
 }
 
+function isValidNextPathFill(path: Fill[], pathFlags: number, fill: Fill): boolean {
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts:127]: We try to compute everything incrementally now so we don't need to continuously go over the path at every step.

File: packages/asset-swapper/src/utils/market_operation_utils/fills.ts [hunk 3]
```diff
@@ -287,3 +297,29 @@ export function getPathAdjustedSlippage(
...
+        return ZERO_AMOUNT;
+    }
+    // Penalize paths that fall short of the entire input amount by a factor of
+    // input / targetInput => (i / t)
+    if (side === MarketOperation.Sell) {
+        // (o / i) * (i / t) => (o / t)
+        return output.div(targetInput);
+    }
+    // (i / o) * (i / t)
+    return input.div(output).times(input.div(targetInput));
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/fills.ts:311]: Bug fix. We were previously just doing `targetInput.div(output)` which made incomplete paths look really impressive, so we would make buy splits that made no sense.

File: packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts [hunk 4]
```diff
@@ -42,11 +52,12 @@ function mixPaths(
     targetInput: BigNumber,
     maxSteps: number,
 ): Fill[] {
-    const _maxSteps = Math.max(maxSteps, 16);
-    let bestPath: Fill[] = pathA;
-    let bestPathInput = getPathSize(pathA, targetInput)[0];
-    let bestPathRate = getPathAdjustedRate(side, pathA, targetInput);
+    const _maxSteps = Math.max(maxSteps, 32);
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts:55]: Faster algo means we can do more minimum steps.

File: packages/asset-swapper/src/utils/market_operation_utils/fills.ts [hunk 5]
```diff
@@ -193,8 +193,12 @@ export function getPathAdjustedSize(path: Fill[], targetInput: BigNumber = POSIT
     for (const fill of path) {
         if (input.plus(fill.input).gte(targetInput)) {
             const di = targetInput.minus(input);
-            input = input.plus(di);
-            output = output.plus(fill.adjustedOutput.times(di.div(fill.input)));
+            if (di.gt(0)) {
+                input = input.plus(di);
+                // Penalty does not get interpolated.
+                const penalty = fill.adjustedOutput.minus(fill.output);
```
Comment (dekz) [inline packages/asset-swapper/src/utils/market_operation_utils/fills.ts:193]: ahhhhhhhh

File: packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts (hunk 1 above)
Comment (dekz) [reply packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts:151]: this was an issue for buys? Looks like we aren't even using `fill.rate` anymore 😂

**Review by dekz — APPROVED** [review]

File: packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts (hunk 1 above)
Comment (dorothy-zbornak) [reply packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts:151]: Issue for everything. I'm removing `fill.rate` and `fill.adjustedRate`.

**Pull request merged**